### PR TITLE
Messages in client

### PR DESCRIPTION
**Pull request by bitPogo**

### CODE REVIEW

File: tests/unit/client/data-access/MessagesRepository.spec.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+describe( 'MessagesRepository', () => {
+	it( 'get mw.message and the messages keys to build a message collection', () => {
+		const keys = [ 'msg1', 'msg2', 'msg3' ];
+		const messagePairs = {
+			msg1: 'Nachricht1',
+			msg2: 'noch ne Nachricht',
+			msg3: 'benachrichtigt',
+		} as any;
+
+		const messages: MwMessage = ( key: string, ...params: string[] ) => new MockMwMessage( messagePairs[ key ] );
```
Comment (jakobw) [inline tests/unit/client/data-access/MessagesRepository.spec.ts]: IMO we don't really need to use the `MockMwMessage` class here. We can create a test implementation here, e.g. like so

```ts
		const messagePairs: { [key: string]: string } = {
			msg1: 'Nachricht1',
			msg2: 'noch ne Nachricht',
			msg3: 'benachrichtigt',
		};

		const messages = ( key: string ) => {
			return { text() { return messagePairs[ key ]; } };
		};
```

File: src/mock-data/MockMwMessage.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+export default class MockMwMessage {
```
Comment (jakobw) [inline src/mock-data/MockMwMessage.ts]: I don't think this class is needed

File: src/mock-data/MockMwMessages.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import MockMwMessage from '@/mock-data/MockMwMessage';
```
[the remark this answers is not in the document]
Comment (bitPogo) [reply src/mock-data/MockMwMessages.ts]: solved by discussion

File: src/client/data-access/MessagesRepository.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import MessagesRepository from '@/common/data-access/MessagesRepository.ts';
+import MessageCollection from '@/datamodel/MessageCollection';
+import { MwMessage } from '@/client/mediawiki/MwWindow';
+
+export default class ClientMessagesRepository extends MessagesRepository {
+
+	private messageInterface: MwMessage;
```
[the remark this answers is not in the document]
Comment (bitPogo) [reply src/client/data-access/MessagesRepository.ts]: done

File: src/mockup-entry.ts [hunk 5]
```diff
@@ -1,13 +1,12 @@
...
-import { StringTMap } from '@/datamodel/LanguageTranslations';
 import entity from './mock-data/data/Q64_data.json';
-import * as languageMap from '@/mock-data/data/en_lang_data.json';
 import * as directionalities from '@/mock-data/data/en_lang_dir_data.json';
 import ImmediatelyInvokingEntityLoadedHookHandler from '@/mock-data/ImmediatelyInvokingEntityLoadedHookHandler';
 import MwWindow from '@/client/mediawiki/MwWindow';
 import getOrEnforceUrlParameter from './mock-data/getOrEnforceUrlParameter';
 
 const language = getOrEnforceUrlParameter( 'language', 'de' );
+import MockupWikibaseContentLanguages from '@/mock-data/MockupWikibaseContentLanguages';
```
[the remark this answers is not in the document]
Comment (bitPogo) [reply src/mockup-entry.ts]: done

File: tests/unit/client/data-access/MessagesRepository.spec.ts (hunk 1 above)
Comment (bitPogo) [reply tests/unit/client/data-access/MessagesRepository.spec.ts]: done

File: src/mock-data/MockMwMessage.ts (hunk 2 above)
Comment (bitPogo) [reply src/mock-data/MockMwMessage.ts]: Done

File: src/common/MessageKeys.ts [hunk 6]
```diff
@@ -0,0 +1,3 @@
+export const MESSAGEKEY_EDIT = 'wikibase-edit';
+
+export const messageKeyList = [MESSAGEKEY_EDIT];
```
[the remark this answers is not in the document]
Comment (bitPogo) [reply src/common/MessageKeys.ts]: done

File: src/client/mediawiki/MwWindow.ts [hunk 7]
```diff
@@ -7,10 +7,17 @@ interface MwConfig {
 
 type MwTitle = new( title: string, namespace: number ) => any;
 
+export interface MwMessageText {
```
Comment (jakobw) [inline src/client/mediawiki/MwWindow.ts]: I find this super confusing, but it's mediawiki's fault. To me, this is the thing that should be called `Message`, and the one below is the service/function that spits out message objects. Naming is hard :[

File: src/components/App.vue [hunk 8]
```diff
@@ -43,6 +44,7 @@ export default class App extends ( Vue as VueConstructor<AppBindings> ) {
 			store.dispatch( `${NS_ENTITY}/${ENTITY_INIT}`, request.entityId ),
 			store.dispatch( `${NS_USER}/${LANGUAGE_PREFERENCE}`, request.language ),
 			store.dispatch( `${NS_LINKS}/${EDIT_LINK_URL_INIT}`, request.editLinkUrl ),
+			store.dispatch( `${NS_MESSAGES}/${MESSAGES_INIT}`, request.language ),
```
Comment (wiese) [inline src/components/App.vue]: I expected this to be done from the [user action](https://github.com/wmde/wikibase-termbox/blob/f1924af4080fc669926219d5612dae07d6ac812d/src/store/user/actions.ts#L13) when its preferred language is set/changed. Objections.

Comment (bitPogo) [reply src/components/App.vue]: done

File: src/client/mediawiki/MwWindow.ts (hunk 7 above)
Comment (bitPogo) [reply src/client/mediawiki/MwWindow.ts]: done

File: src/store/user/actions.ts [hunk 9]
```diff
@@ -11,6 +15,7 @@ export const actions: ActionTree<User, any> = {
 		context.commit( LANGUAGE_INIT, language );
 
 		// this would be a good place to dispatch a message translation update, too
```
Comment (wiese) [inline src/store/user/actions.ts]:
```suggestion
```

File: src/store/user/actions.ts [hunk 10]
```diff
@@ -11,6 +15,7 @@ export const actions: ActionTree<User, any> = {
...
+		context.dispatch( `${NS_MESSAGES}/${MESSAGES_INIT}`, language, { root: true } );
```
Comment (wiese) [inline src/store/user/actions.ts]:
```suggestion
		return Promise.all( [
			context.dispatch( `${NS_MESSAGES}/${MESSAGES_INIT}`, language, { root: true } ),
			context.dispatch( `${NS_LANGUAGE}/${ENSURE_AVAILABLE_IN_LANGUAGE}`, language, { root: true } );
		] );
```

File: src/store/user/actions.ts [hunk 11]
```diff
@@ -11,6 +15,7 @@ export const actions: ActionTree<User, any> = {
...
 		return context.dispatch( `${NS_LANGUAGE}/${ENSURE_AVAILABLE_IN_LANGUAGE}`, language, { root: true } );
```
Comment (wiese) [inline src/store/user/actions.ts]:
```suggestion
```

File: src/components/App.vue [hunk 12]
```diff
@@ -43,6 +43,7 @@ export default class App extends ( Vue as VueConstructor<AppBindings> ) {
 			store.dispatch( `${NS_ENTITY}/${ENTITY_INIT}`, request.entityId ),
 			store.dispatch( `${NS_USER}/${LANGUAGE_PREFERENCE}`, request.language ),
 			store.dispatch( `${NS_LINKS}/${EDIT_LINK_URL_INIT}`, request.editLinkUrl ),
+			/* Messages are dispatched in src/store/user/actions.ts */
```
Comment (wiese) [inline src/components/App.vue]: So are language translations. I'm not strongly objecting that comment but wonder if comments should strive to replace a general overview of the code base.

Comment (jakobw) [reply src/components/App.vue]: I agree and removed the comment for now!

File: src/store/user/actions.ts (hunk 9 above)
Comment (jakobw) [reply src/store/user/actions.ts]: Thanks! Done.

**Pull request merged**